File: 297-serialize-and-deserialize-binary-tree.java
public class Codec {
    private static final String SPLITTER = ",";
    private static final String NULL_SYMBOL = "X";

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        StringBuilder sb = new StringBuilder();
        buildString(root, sb);
        return sb.toString();
    }

    private void buildString(TreeNode root, StringBuilder sb) {
        if(root == null) {
            sb.append(NULL_SYMBOL).append(SPLITTER);
        }
        else {
            sb.append(root.val).append(SPLITTER);
            buildString(root.left, sb);
            buildString(root.right, sb);
        }
    }
    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        Queue<String> nodes = new LinkedList<>();
        nodes.addAll(Arrays.asList(data.split(SPLITTER)));
        return buildTree(nodes);
    }
             
    private TreeNode buildTree(Queue<String> nodes) {
        String val = nodes.remove();
        if(NULL_SYMBOL.equals(val)) {
            return null;
        }
        
        TreeNode root = new TreeNode(Integer.valueOf(val));
        root.left = buildTree(nodes);
        root.right = buildTree(nodes);
        
        return root;
    }
}

File: 297_Serialize_and_Deserialize_Binary_Tree.py
I use preorder DFS to serialize the tree.
Every node is recorded as its value, and I use "x" for null pointers.
This sequence uniquely describes the tree because preorder always visits: node → left → right.

To deserialize, I read the values in order using an iterator.
Whenever I see "x" I return None; otherwise I create a node and recursively build its left and right subtrees.
Because serialize and deserialize follow the same preorder structure, the original tree is reconstructed exactly.

class Codec:

    def serialize(self, root):
        values = []

        def dfs(node):
            if not node:
                values.append("x")
                return
            values.append(str(node.val))
            dfs(node.left)
            dfs(node.right)

        dfs(root)
        return ",".join(values)

    def deserialize(self, data):
        values = iter(data.split(","))

        def build_subtree():
            val = next(values)
            if val == "x":
                return None
            node = TreeNode(int(val))
            node.left = build_subtree()
            node.right = build_subtree()
            return node

        return build_subtree()
